BidirectionalActions implemented

Manthan-R-Sheth opened the pull request:
This is an initial approach for implementing BidirectionalActions interface.

Manthan-R-Sheth committed:
BidirectionalActions implemented and BidirectionalSearch API updated and independent calls handled

Manthan-R-Sheth commented:
@ctjoreilly  
This is with regards to #264. It is failing BidirectionalSearchTest as the API is now updated to support ```BidirectionalActions```. 
It is an initial approach but i suppose it gives desired results except for corner cases. By corner cases, I mean  the node that is found to be existing in other frontier is not added to the list. I can use manual addition though.
I have maintained a HashMap to get the parent of the child node in the other frontier(if it exists). ```newChildNode()``` assigns the new parent of the present frontier and so I had to use the HashMap.
Kindly give your suggestions.

Manthan-R-Sheth commented:
@ctjoreilly 
All functions work as expected. Review it now.

ctjoreilly commented on core/src/main/java/aima/core/search/api/BidirectionalActions.java:
```diff
@@ -0,0 +1,13 @@
+
+/**
+ * Created by manthan on 13/3/17.
+ */
```
@author manthan

ctjoreilly commented on core/src/main/java/aima/core/search/api/BidirectionalActions.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Created by manthan on 13/3/17.
+ */
+public interface BidirectionalActions<A> {
```
Could you add JavaDoc describing the methods.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -5,11 +5,7 @@
-import java.util.Queue;
-import java.util.Set;
-import java.util.HashSet;
+import java.util.*;
```
Could we list just the classes imported and avoid .*.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -22,68 +18,87 @@
-        Set<S> exploredFront = newExploredSet();
-        Set<S> exploredBack = newExploredSet();
+        HashMap<S, Node<A, S>> exploredFront = newExploredMap();
+        HashMap<S, Node<A, S>> exploredBack = newExploredMap();
```
These should be 'Map<..>' and not 'HashMap<...>' - i.e. avoid restricting to a specific implementation.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -22,68 +18,87 @@
-        exploredFront.add(node.state());
-        exploredBack.add(revNode.state());
+        exploredFront.put(node.state(), null);
+        exploredBack.put(revNode.state(), null);
```
newExploredMap() could take an state argument so that these line would not need to be repeated. e.g.:
newExploredMap(node.state())

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -22,68 +18,87 @@
 
             if (front.isEmpty() || back.isEmpty()) {
-                return failure();
+                this.previousMeetingOfTwoFrontiers = null;
```
I think you should still use a failure() method here, which implements its own interpretation of how failure is represented for this algorithm (all empty lists returned should suffice). I realize you are doing this due to implementing BidirectionalActions interface but this should be cleaned up.

Manthan-R-Sheth replied:
@ctjoreilly, By the present implementation, failure() should be called whenever there is an empty list condition.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -22,68 +18,87 @@
-                List<A> sol = solution(solutionNode);
-                Collections.reverse(sol);
-                return sol;
+                this.meetingOfTwoFrontiers = solutionNode;
```
I think you could replace these two conditional blocks with a method call something like:

if (isSolution(pathExistsBidirectional(front, exploredFront, exploredBack, originalProblem), true)) {
return bidirectionalSearch;
}

where the 2nd argument represents this.fromFront's value to assign.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -107,11 +122,65 @@
-    public Set<S> newExploredSet() {
-        return new HashSet<>();
+    public HashMap<S, Node<A, S>> newExploredMap() {
+        return new HashMap<>();
```
public Map<...

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -22,68 +18,87 @@
  * @author manthan
  */
-public class BidirectionalSearch<A, S> implements SearchForActionsBidirectionallyFunction<A, S> {
+public class BidirectionalSearch<A, S> implements SearchForActionsBidirectionallyFunction<A, S>, BidirectionalActions<A> {
```
Not a good idea having the search algorithm implement BidirectionalActions<A>. What if apply is called twice with two different arguments? Ideally, just create a simple:

aima.core.search.basic.support.BasicBidirectionalActions

class and use it.

ctjoreilly reviewed (changes requested):
Getting close, could you review feedback and make appropriate changes. The key change is not to have the search algorithm implement the BidirectionalActions interface. Once updated could you squash/rebase as necessary so that I can apply patch and try out before merging.

Thanks

Ciaran

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicBidirectionalActions.java:
```diff
@@ -0,0 +1,86 @@
+    private Node<A, S> meetingOfTwoFrontiers;
+    private boolean fromFront;
+    protected SearchController<A, S> searchController = new BasicSearchController<>();
+
```
This class is too complex for a basic support class as well as being too coupled to the current BidirectionalSearch's implementation approach. The constructor for this class should only need to take the fromInitialListPartList and the fromGoalStatePartList and to derive the other two lists from them by joining/sorting appropriately. The other logic should migrate back into BidirectionalSearch as supporting routines.

Manthan-R-Sheth replied:
Ya I realise that.
Am on it.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -1,15 +1,13 @@
 package aima.core.search.basic.uninformed;
 
 import aima.core.search.api.*;
```
Expand api.* to classes actually required.

ctjoreilly reviewed (changes requested):
Almost there, just a bit more refactoring and we should be good. Thanks.

Manthan-R-Sheth commented:
@ctjoreilly 
Updated the PR.

Manthan-R-Sheth commented:
@ctjoreilly
Updated the PR and squashed the commits.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicBidirectionalActions.java:
```diff
@@ -0,0 +1,50 @@
+        fromInitialStateToGoalStateList.addAll(fromInitialStatePart());
+        fromInitialStateToGoalStateList.addAll(sol);
+        return fromInitialStateToGoalStateList;
+    }
```
What would be returned if this method was called twice in a row. I think it would double the size of the list of actions returned. Setting up fromInitialStateToGoalStateList and fromGoalStateToInitialStateList should only need to happen once in the constructor or a special/private init method.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicBidirectionalActions.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public List<A> fromInitialStateToGoalState() {
+        List<A> sol = fromGoalStatePart();
+        Collections.reverse(sol);
```
You've actually reversed the 'fromGoalStatePartList' corrupting for subsequent calls to fromGoalStatePart(). You need to do:

List<A> reversedFromGoalStatePart = new ArrayList<>(fromGoal...);

ctjoreilly reviewed (changes requested):
Very close but please review my comments related to BasicBidirectionalActions, address them and write a BasicBidirectionalActionsTest class demonstrating it works correctly when the API method are called several times in a row - i.e. these API calls should be idempotent.

Manthan-R-Sheth commented:
@ctjoreilly 
I have fixed the problem by an ```init()``` that initialises ```fromInitialStateToGoalStateList ``` and ```fromGoalStateToInitialStateList```. Supporting tests are also added.

ctjoreilly commented:
Can you please rebase/squash as need so I can apply as a patch to try out (currently does not work).

Manthan-R-Sheth commented:
@ctjoreilly 
It should work now.
